Feed API

gbuisson opened the pull request:
> **Epic** https://github.com/threatgrid/iroh/issues/2417
> Close https://github.com/threatgrid/iroh/issues/3085

This PR adds support for the Indicator Feed API including feed Views


<a name="qa">[§](#qa)</a> QA
============================

## Test Feed Crud

A new set of routes is available under /feed (https://intel.int.iroh.site/ctia/index.html#/Feed),
they support the regular CRUD and search operations common to all the other entities

the specificity of this entity is that is has 3 other calculated fields: `secret`, `feed_view_url` and `feed_view_url_csv`, those fields are created at POST and shouldn't be updatable which means that if you try to use the update route specifying them, thy should be ignored. 

## Test Feed Views

The `feed_view_url` and `feed_view_url_csv` fields reference views available without authentication.

### Test Setup

- Create an Indicator and some Judgements, link them with Relationships
- Create a Feed and set its output type as `observables`, additionally set its `lifetime` to a valid time range.

### View tests steps

- fetch `feed_view_url`, it should return all `Judgement` `Observables` without any duplicate as the content type of your choice specified by the `accept` header.
- fetch `feed_view_url_csv`,  it should return all `Judgement` `Observables` values  without any duplicate as a CSV string

- update the Feed, setting its `output` field to `judgements`

- fetch `feed_view_url`, it should return all `Judgements`  without any duplicate as the content type of your choice specified by the `accept` header.
- fetch  fetch `feed_view_url_csv`,  it should return all `Judgements`  without any duplicate as a CSV string

### Miscellaneous steps

- update the Feed, set its lifetime to something outdated, the views should return a 404
- fetch a View altering the `s` query parameter, you should get an `Unauthorized` response status
- link the Target Indicator to a non existing Judgement, the Feed View should still work ignoring it
- link the Target Indicator to a Judgement you don't own, the Feed view should still work ignoring it

<a name="ops">[§](#ops)</a> Ops
===============================

- a new `feed` index should be created

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
public: New CTIA Feed API.
```

msprunck commented on src/ctia/entity/feed/schemas.clj:
```diff
@@ -0,0 +1,112 @@
+             :groups (or (:groups prev-object) groups)
+             :secret secret
+             :feed_view_url_csv feed_view_url_csv
+             :feed_view_url feed_view_url
```
I think that `secret`, `feed_view_url_csv` and `feed_view_url` must be encrypted

msprunck commented on src/ctia/entity/feed/schemas.clj:
```diff
@@ -0,0 +1,112 @@
+             :timestamp (or (:timestamp new-object) now)
+             :tlp (:tlp new-object
+                        (:tlp prev-object (properties-default-tlp)))}
+            (when (contains-key? Feed :valid_time)
```
```suggestion
            (when (contains-key? Feed :lifetime)
```

msprunck commented on src/ctia/entity/feed/schemas.clj:
```diff
@@ -0,0 +1,112 @@
+                        (:tlp prev-object (properties-default-tlp)))}
+            (when (contains-key? Feed :valid_time)
+              (make-valid-time
+               (:valid_time prev-object)
```
```suggestion
               (:lifetime prev-object)
```

msprunck commented on src/ctia/entity/feed/schemas.clj:
```diff
@@ -0,0 +1,112 @@
+            (when (contains-key? Feed :valid_time)
+              (make-valid-time
+               (:valid_time prev-object)
+               (:valid_time new-object)
```
```suggestion
               (:lifetime new-object)
```

msprunck commented on src/ctia/entity/feed/schemas.clj:
```diff
@@ -0,0 +1,112 @@
+
+(s/defschema PartialFeedList [PartialFeed])
+
+(s/defn realize-feed :- StoredFeed
```
Could you ensure in the realize function that the `indicator_id` is provided if the `feed_type` is `:indicator`?

msprunck replied:
Where did you add this check?

msprunck commented on src/ctia/entity/feed.clj:
```diff
@@ -0,0 +1,239 @@
+                            feed-identity
+                            {:fields [:source_ref]})
+                           (map :source_ref)
+                           (remove nil?)
```
You can replace these two lines by `(keep :source_ref)`

msprunck commented on test/ctia/entity/feed_test.clj:
```diff
@@ -0,0 +1,377 @@
+                                              (string/join ""))
+          response (client/get feed-view-url
+                               {:as :json
+                                :headers {"Authorization" "45c1f5e3f05d0"}})
```
You should not define an authorisation header, I understood that this route should only require the secret

msprunck commented on test/ctia/entity/feed_test.clj:
```diff
@@ -0,0 +1,377 @@
+                               {:as :json
+                                :headers {"Authorization" "45c1f5e3f05d0"}})
+          response-csv (client/get feed-view-url-csv
+                                   {:headers {"Authorization" "45c1f5e3f05d0"}})
```
Same here

msprunck commented on src/ctia/lib/csv.clj:
```diff
@@ -0,0 +1,89 @@
+                  (when include-headers?
+                    (string/join "," (map #(field-protect (name %)) ks)))
+                  (for [m flattened-ms]
+                    (string/join "," (map #(field-protect (get m % "")) ks)))))))
```
Could you test this function?

msprunck commented on src/ctia/lib/csv.clj:
```diff
@@ -0,0 +1,89 @@
+                       {:compojure.api.meta/serializable? false})]
+    (update-in response [:headers]
+               #(into % (csv-http-headers
+                         (uri->filename uri))))))
```
Could you test this function?

msprunck commented on src/ctia/lib/csv.clj:
```diff
@@ -0,0 +1,89 @@
+(defn deep-flatten-map
+  ([m] (deep-flatten-map "" m))
+  ([prefix m]
+   (into {} (deep-flatten-map-as-couples prefix m))))
```
Could you test this function?

msprunck commented on src/ctia/entity/feed.clj:
```diff
@@ -0,0 +1,239 @@
+           (map #(select-keys % [:value]) data)))
+
+(defn render-headers? [output]
+  (not= :observables output))
```
I don't like the fact that headers are rendered or not based on the output type. Could you add a query param `render-headers`?

gbuisson replied:
I think we shouldn't add query parameters for this route because we may switch to generated views if we get performance issues.

ereteog commented on src/ctia/stores/es/crud.clj:
```diff
@@ -179,7 +178,8 @@
                          coerce!)]
-        (if (allow-read? doc ident)
+        (if (or (not (restricted-read? ident))
+                (allow-read? doc ident))
```
Could you add some tests related to `:authorized-anonymous` field in the identity?

ereteog commented on src/ctia/stores/es/crud.clj:
```diff
@@ -282,25 +282,27 @@
-                             (find-restriction-query-part ident))
+      (let [filter-val (cond-> (q/prepare-terms all-of)
+                         (restricted-read? ident)
+                         (conj (find-restriction-query-part ident)))
```
same thing for find, could you add some tests related to `:authorized-anonymous` field in the identity?

gbuisson commented on src/ctia/entity/feed.clj:
```diff
@@ -0,0 +1,239 @@
+
+(def feed-view-routes
+  (routes
+   (GET "/:id/view.csv" []
```
we don't do it that way because this view is intended to people who don't know how to use HTTP headers.

gbuisson commented:
Based on Feedback I added:

- Feed secrets encryption with a new CTIA encryption service (tested)
- Augmented the AC Tests for the `:authoryzed-anonymous` feature
- Tests for the CSV rendering part
- A description of the feature and how to use it in the readme

I also corrected all the small changes you required.
Let me know if you want something more precisely.

hshahman commented:
Verified in rel-v1.41
